sdk/src/main/java/com/horizen/transaction/RegularTransaction.java

```diff
@@ -92,17 +94,17 @@ public Proof boxKey() {
     public synchronized List<NoncedBox<Proposition>> newBoxes() {
         if(newBoxes == null) {
             newBoxes = new ArrayList<>();
-            int boxIndex = 0;
-            for (int i = 0; i < regularOutputs.size(); i++, boxIndex++) {
-                NoncedBox box = new RegularBox(regularOutputs.get(i).getKey(), getNewBoxNonce(regularOutputs.get(i).getKey(), boxIndex),
-                        regularOutputs.get(i).getValue());
-                newBoxes.add(box);
-            }
-            for (int i = 0; i < withdrawalOutputs.size(); i++, boxIndex++) {
-                NoncedBox box = new WithdrawalRequestBox(withdrawalOutputs.get(i).getKey(),
-                        getNewBoxNonce(withdrawalOutputs.get(i).getKey(), boxIndex),
-                        withdrawalOutputs.get(i).getValue());
-                newBoxes.add(box);
+            for (int i = 0; i < outputs.size(); i++) {
+                long nonce = getNewBoxNonce(outputs.get(i).proposition(), i);
+                BoxData boxData = outputs.get(i);
+                if(boxData instanceof RegularBoxData) {
```

Comment:
why not to add into BoxData method for creation appropriate box according boxData?

Reply:
Discussion related to hard bounding of box and boxdata, possibility to create box by boxdata.

Reply:
create boxes from data by BoxData.createAppropriateBox() ?